### PR TITLE
Enabled firebase to be called as function, thus enabling patterns where Firebase ref's are stored in root component

### PR DESCRIPTION
**Pull request by DaBs**
This pull request will enable using the firebase option in the same way as the data option, by calling it as a function and returning an object. This allows for coding patterns where you originally define the Firebase references as data in the root component, but want to use the later on in child components, e.g. 
```javascript
export default {
    ...,
    firebase() {
        return {
            someArray: this.$root.database.ref('someArray')
        }
    },
    ...
}
```

### CODE REVIEW

**Comment by jwngr** [comment]
Can you just use `firebase.database.ref('someArray')` instead `this.$root.database.ref('someArray')`? That should make it so that you don't need to have a function here and it's even fewer characters.

**Comment by DaBs** [comment]
Hm, sure, albeit this means you have to (with ES6 modules atleast) import the firebase module everywhere you want to use it, which is then a lot more characters than the above mentioned (if that's what we're measuring). Furthermore, I don't understand why this would not be intended behaviour, seeing as most other VueJS options are capable of being defined this way.

**Comment by jwngr** [comment]
The fewer characters was a bit tongue-in-cheek. I'll leave it to Evan to comment on whether or not this makes sense from a Vue perspective. From a Firebase perspective though, I think it would be a bit redundant.

**Comment by DaBs** [comment]
Heh, well aware, I was just throwing it right back at you in the same manner. 😛  No offence taken. And I realise that this is sort of opposite of what the idea behind the way Firebase wants to both store its own state and references, but attempting to utilize this in a VueJS component context, it made little sense why it wouldn't behave like the `data() { ... }` option.

**Comment by jwngr** [comment]
Haha makes sense 😄 I am not familiar enough with VueJS to say whether or not this is a good idea. It certainly seems like it takes very little code to support and I guess I don't see a reason not to include it, but I'll leave that to Evan to decide.

Thanks for the contribution by the way! I'm always happy to see the community getting involved!

**Comment by DaBs** [comment]
No problem, I'm just happy that two of my favourite stack choices right now have a well-functioning module and I'll gladly contribute to ensure that it continues to be well-functioning 😄

**Comment by posva** [comment]
@DaBs Can you add a test, please?
Also, I think you can remove the compiled version. That seems to be done in a separate commit

**Comment by DaBs** [comment]
@posva I'm not sure if this covers what you wanted from a test case, but it tests the basic functionality of whether or not you can use the firebase option as a function.

File: tests/vuefire.spec.js [hunk 1]
```diff
@@ -12,9 +12,11 @@ var firebaseApp = Firebase.initializeApp({
 
 describe('VueFire', function () {
   var firebaseRef
+  var firebaseDb
 
   beforeEach(function (done) {
-    firebaseRef = firebaseApp.database().ref()
+    firebaseDb = firebaseApp.database()
```
Comment (posva) [inline tests/vuefire.spec.js]: This can be moved inside the `it` function as it's only used there

**Comment by posva** [comment]
@DaBs Yes, thank you.
About the compiled version, you'll have to rebase and remove the modification in the build. You have to commit everything but the dist files.

File: tests/vuefire.spec.js [hunk 2]
```diff
@@ -25,6 +27,36 @@ describe('VueFire', function () {
     })
   })
 
+  describe('is callable as Function', function () {
```
Comment (posva) [inline tests/vuefire.spec.js]: The title is missing something about __option__, which is what you're actually changing. I'd name it something like _support Function options_

File: tests/vuefire.spec.js [hunk 3]
```diff
@@ -25,6 +27,36 @@ describe('VueFire', function () {
...
+    it('returns correct ref on function call', function (done) {
+      var ChildComponent = Vue.extend({
+        name: 'ChildComponent',
+        firebase: function () {
```
Comment (posva) [inline tests/vuefire.spec.js]: Can't you just use this inside the Vue instance? Wouldn't that be easier?

File: tests/vuefire.spec.js [hunk 4]
```diff
@@ -25,6 +27,36 @@ describe('VueFire', function () {
...
+        },
+        template: '<div>test</div>'
+      })
+      new Vue({
+        data: function () {
+          return {
+            database: firebaseDb
+          }
+        },
+        ready: function () {
```
Comment (posva) [inline tests/vuefire.spec.js]: The `ready` hook is unnecessary

File: tests/vuefire.spec.js [hunk 5]
```diff
@@ -25,6 +27,36 @@ describe('VueFire', function () {
...
+          expect('poop').to.equal('notpoop')
+        },
+        components: {
+          'child-component': ChildComponent
+        },
+        template: '<div><child-component></child-component></div>'
+      }).$mount()
+      Vue.nextTick(function () {
```
Comment (posva) [inline tests/vuefire.spec.js]: This is also unnecessary

Comment (DaBs) [inline tests/vuefire.spec.js]: Damn, I seem to not have removed my initial attempts at getting the test to work. I'll fix that ASAP.

File: tests/vuefire.spec.js [hunk 6]
```diff
@@ -25,6 +25,30 @@ describe('VueFire', function () {
...
+      new Vue({
+        data: function () {
+          return {
+            database: firebaseDb
+          }
+        },
+        components: {
+          'child-component': Vue.extend({
+            name: 'ChildComponent',
+            firebase: function () {
```
Comment (posva) [inline tests/vuefire.spec.js]: Do you actually need to have a child component?

Comment (DaBs) [inline tests/vuefire.spec.js]: It could easily be simplified to simply test the call of the function. I'll change immediately.

**Comment by DaBs** [comment]
@posva Does the test satisfy the standard? I'm sorry it took so long to get a proper test, but writing unit tests in chai is relatively new for me :smile:

**Comment by posva** [comment]
@DaBs It's ok, I actually feel being pushy 😓 
The unit test is much better that way. It'd be perfect if you can use a spy and check it gets called only once after the `$mount`
I'd also remove the regex on the throw. If it throws anything it can be considered as a failure

**Comment by DaBs** [comment]
@posva Sorry for the delay, I was busy most of the week. The latest commit should have added both a spy and removed the regex 😄

File: src/vuefire.js [hunk 7]
```diff
@@ -198,6 +198,7 @@ function ensureRefs (vm) {
 var init = function () {
   var bindings = this.$options.firebase
   if (!bindings) return
```
Comment (posva) [inline src/vuefire.js]: I think it makes sense to move this one line below, so if the function is returning nothing we can skip too

**Comment by posva** [comment]
@DaBs That's okay 😄

File: src/vuefire.js [hunk 8]
```diff
@@ -196,8 +196,9 @@ function ensureRefs (vm) {
 }
 
 var init = function () {
+  if (!this.$options.firebase) return
```
Comment (posva) [inline src/vuefire.js]: I wasn't clear about this, sorry. I was thinking about this: 

``` js
var bindings = this.$options.firebase
if (typeof bindings === 'function') bindings = bindings.call(this)
if (!bindings) return
```

**Comment by DaBs** [comment]
What is status on this PR? :smile:

**Comment by posva** [comment]
@DaBs Sorry for not giving you any updates! Can you remove the modifications on the dist folder, so we only keep the src and test modifications?

**Comment by DaBs** [comment]
@posva Ah, damn, that reoccured. Will do ASAP.

**Comment by DaBs** [comment]
@posva This should do it 😄

**Comment by posva** [comment]
@DaBs Thank you very much! I'm merging this and adding some minor changes to the test!

**Pull request merged**